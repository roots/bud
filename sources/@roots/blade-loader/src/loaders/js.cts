import {extract} from './index.cjs'

export const extension = `js`
export const pattern =
  /@module\(['"]js['"]\)(?<content>[\s\S]*?)@endmodule/g

export default (source: string) => extract(source, pattern)
