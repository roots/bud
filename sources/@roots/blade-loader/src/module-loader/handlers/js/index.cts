import {extract} from '@roots/blade-loader/module-loader/handlers'

export const extension = `js`
export const pattern =
  /@module\(['"]js['"]\)(?<content>[\s\S]*?)@endmodule/g

export default (source: string) => extract(source, pattern)
