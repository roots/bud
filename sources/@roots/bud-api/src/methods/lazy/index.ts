import type {Bud} from '@roots/bud-framework'
import type {LazyCompilationOptions} from '@roots/bud-framework'

export type Parameters = [boolean | LazyCompilationOptions]

/**
 * Lazy function interface
 */
export interface lazy {
  (...parameters: Parameters): Bud
}

/**
 * Enables lazy compilation of built assets.
 *
 * @example
 * Enable:
 *
 * ```js
 * bud.lazy()
 * ```
 *
 * @example
 * Explicitly disable:
 *
 * ```js
 * bud.lazy(false)
 * ```
 *
 * @example
 * Explicitly enable:
 *
 * ```js
 * bud.lazy(true)
 * ```
 *
 * @example
 * Use with object options:
 *
 * ```js
 * bud.lazy({
 *   entries: false,
 *   imports: true,
 * })
 * ```
 */
export const lazy: lazy = function (this: Bud, value = true) {
  if (typeof value === `boolean`) {
    this.hooks.on(`build.experiments`, experiments => ({
      ...(experiments ?? {}),
      lazyCompilation: value,
    }))
    return this
  }

  this.hooks.on(`build.experiments`, experiments => {
    if (
      typeof experiments?.lazyCompilation === `boolean` ||
      !experiments?.lazyCompilation
    )
      return {...(experiments ?? {}), lazyCompilation: value}

    experiments.lazyCompilation = {
      ...experiments.lazyCompilation,
      ...value,
    }

    return experiments
  })

  return this
}
