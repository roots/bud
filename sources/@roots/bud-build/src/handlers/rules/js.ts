import type {Factory} from '../index.js'

export const js: Factory = ({filter, makeRule, path}) =>
  makeRule()
    .setTest(filter(`pattern.js`))
    .setInclude([() => path(`@src`)])
    .setUse([])
