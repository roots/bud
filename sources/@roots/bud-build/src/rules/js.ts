import type {Factory} from '@roots/bud-build/registry'

const js: Factory = async ({filter, makeRule, path}) =>
  makeRule()
    .setTest(filter(`pattern.js`))
    .setInclude([() => path(`@src`)])
    .setResolve({
      fullySpecified: false,
    })

export {js as default}
