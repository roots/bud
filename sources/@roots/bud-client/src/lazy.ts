/**
 * Lazy import helper
 *
 * @remarks
 * Callback function may be async or sync
 *
 * @param onReady - callback function
 * @returns void
 */
interface lazy {
  <T = any>(
    module: Promise<T>,
    handler: (module: T) => Promise<unknown> | unknown,
    errorHandler?: (error: unknown) => unknown,
  ): Promise<unknown>
}

/**
 * Default error handler
 *
 * @throws Error
 */
const defaultErrorHandler = (error: unknown) => {
  throw error
}

const lazy: lazy = async function lazy<T = any>(
  module: Promise<T>,
  onImport: (module: T) => Promise<unknown> | unknown,
  onError?: (error: unknown) => unknown,
) {
  try {
    const request = await module
    if (!request) throw new Error(`module not found: ${module}`)
    const result = await onImport(request)
    return result
  } catch (error: unknown) {
    const errorFn = onError ? onError : defaultErrorHandler
    errorFn(error)
  }
}

export default lazy
