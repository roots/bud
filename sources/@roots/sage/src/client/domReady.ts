/**
 * Calls given function once document has loaded.
 *
 * @param onReady - callback function
 * @returns void
 *
 * @public
 */
interface domReady {
  (onReady: () => void): void
}

const domReady: domReady = onReady => {
  window.requestAnimationFrame(function check() {
    document.body ? onReady() : window.requestAnimationFrame(check)
  })
}

export default domReady
