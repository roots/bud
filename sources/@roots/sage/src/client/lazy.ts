/**
 * Lazy import helper
 *
 * @remarks
 * Callback function may be async or sync
 *
 * @param onReady - callback function
 * @returns void
 */
interface lazy {
  <T = any>(
    module: Promise<{default: T}>,
    handler: (module: T) => unknown | Promise<unknown>,
    errorHandler?: (err: Error) => unknown,
  ): Promise<unknown>
}

/**
 * Default error handler
 *
 * @throws Error
 * @public
 */
const defaultErrorHandler = (err: string) => {
  throw new Error(err)
}

const lazy: lazy = async function lazy<T = any>(
  module: Promise<{default: T}>,
  handler: (module: T) => unknown | Promise<unknown>,
  errorHandler?: (err: Error) => unknown,
) {
  try {
    const {default: request} = await module
    return await handler(request)
  } catch (err) {
    const handle = errorHandler ? errorHandler : defaultErrorHandler
    handle(err)
  }
}

export default lazy
