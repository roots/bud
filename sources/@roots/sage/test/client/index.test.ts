import {describe, expect, it} from 'vitest'

import domReady from '../../src/client/dom-ready.js'
import * as client from '../../src/client/index.js'
import lazy from '../../src/client/lazy.js'

describe(`@roots/sage`, () => {
  it(`@roots/sage/client`, () => {
    expect(client).toMatchSnapshot()
  })

  it(`@roots/sage/client/dom-ready`, () => {
    expect(client.domReady).toBe(domReady)
  })

  it(`@roots/sage/client/lazy`, () => {
    expect(client.lazy).toBe(lazy)
  })
})
