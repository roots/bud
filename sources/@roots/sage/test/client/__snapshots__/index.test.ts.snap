// Vitest Snapshot v1

exports[`@roots/sage > @roots/sage/client 1`] = `
{
  "domReady": [Function],
  "lazy": [Function],
}
`;
